[Build] Make test cluster plugin configuration cache compatible

breskeby opened the pull request:
This adds infrastructure to make the legacy test cluster plugin and
the legacy test cluster based test plugins generally configuration cache compatible.

breskeby committed:
Fix running cluster lookup

- after loaded from configuration cache
- in multiproject running parallel scenario

breskeby committed:
[Gradle] Make test cluster plugin configuration cache compatible

This adds infrastructure to make the legacy test cluster plugin and
the legacy test cluster based test plugins generally configuration cache compatible.

elasticsearchmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

breskeby committed:
Fix license header

breskeby committed:
Remove debugging

breskeby committed:
Fix cc compatibility for AbstractYamlRestCompatTestPlugin

breskeby committed:
fix imports

breskeby committed:
Some cleanup

breskeby committed:
More cleanup

breskeby commented:
also covered https://github.com/elastic/elasticsearch/pull/110362

breskeby committed:
Fix AntFixture configuration cache compatibility

We need to resolve the address and port from a fixture as ValueSource

breskeby committed:
Cleanup build script

breskeby commented on build-tools-internal/src/main/groovy/org/elasticsearch/gradle/internal/test/AntFixture.groovy:
```diff
@@ -10,22 +10,37 @@
 package org.elasticsearch.gradle.internal.test
 
 import org.elasticsearch.gradle.OS
```
Though we want to ultimately get rid of AntFixture, I did some tweaks here to get `plugins/discovery-ec2/qa/amazon-ec2` configuration cache compatible as part of this.

breskeby commented:
Example run of legacy plugin based integration tests: https://gradle-enterprise.elastic.co/s/6cyrhupfxc3ho

breskeby committed:
Remove more debugging

jozala commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/AbstractYamlRestCompatTestPlugin.java:
```diff
@@ -244,10 +245,11 @@ public void apply(Project project) {
             // Use test runner and classpath from "normal" yaml source set
+            FileCollection sourceSetOutput = yamlCompatTestSourceSet.getOutput();
             testTask.setTestClassesDirs(
                 yamlTestSourceSet.getOutput().getClassesDirs().plus(yamlCompatTestSourceSet.getOutput().getClassesDirs())
```
NIT: This `yamlCompatTestSourceSet.getOutput()` could also be replaced with `sourceSetOutput` since the same was done for readability below.

jozala reviewed (changes requested):
I've left a few comments for some minor improvements.

breskeby committed:
Apply review feedback

- remove unsed code
- fail early when nodes are somehow added to a testcluster after useCluster

breskeby commented on x-pack/plugin/sql/qa/jdbc/security/build.gradle:
```diff
@@ -46,6 +53,17 @@ subprojects {
+      project.gradle.sharedServices,
+      TestClustersPlugin.REGISTRY_SERVICE_NAME
+    )
+    def clusterInfo = project.getProviders().of(TestClusterValueSource.class) {
+      it.parameters.path.set(clusterPath)
+      it.parameters.clusterName.set("javaRestTest")
+      it.parameters.service = serviceProvider
+    }
```
fixed

breskeby commented on build-tools/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersAware.java:
```diff
@@ -34,11 +35,17 @@ default void useCluster(ElasticsearchCluster cluster) {
-        cluster.getNodes()
-            .all(node -> node.getDistributions().forEach(distro -> dependsOn(getProject().provider(() -> distro.maybeFreeze()))));
-        dependsOn(cluster.getPluginAndModuleConfigurations());
+        if (cluster.getName().equals(getName())) {
+            for (ElasticsearchNode node : cluster.getNodes()) {
+                for (ElasticsearchDistribution distro : node.getDistributions()) {
+                    ElasticsearchDistribution frozenDistro = distro.maybeFreeze();
+                    dependsOn(frozenDistro);
+                }
+            }
+            dependsOn(cluster.getPluginAndModuleConfigurations());
+        }
```
`numberOfNodes` is still used across the build in the build scripts. idea is bad in finding those references automatically for you in build scripts unfortunately (especially with the gradle dsl).
The numberOfNodes property should  not be changed after we have triggered the useCluster here. I tweaked this to fail in case this does happen somehow. 
Using a domain object container for ElasticSearchCluster#nodes was probably not required in the first place. It just had some convenience to rely on less configuration ordering in the TestCluster configurations across the build

jozala reviewed (approved):
LGTM. Thanks for explanations.

breskeby committed:
do not freeze cluster setup early

breskeby committed:
Fix mixed cluster tests

breskeby committed:
Fix bwc tests for qa/mixed-cluster

breskeby committed:
Merge branch 'main' into test-cluster-cc-compatible

breskeby commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119365">](https://github.com/elastic/elasticsearch/pull/119365)|
|✅|8.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119366">](https://github.com/elastic/elasticsearch/pull/119366)|
|✅|8.16|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119367">](https://github.com/elastic/elasticsearch/pull/119367)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->
